Revert fdc3.timeRange and fdc3.interaction oneOf back to anyOf and add example validation tests

bingenito opened the pull request:
## Describe your change

* Add unit tests to the fdc3-context package for validating context examples are valid schema.
* Revert schema of `fdc3.timeRange` context type back to use anyOf in place of oneOf for the `startTime` and `endTime` property combinations.  This will allow existence of one of either, or both, and pass schema validation.  When defined with oneOf, validation would fail due to multiple entries being valid and it could not identify which to apply.
* Revert schema of `fdc3.interaction` context type back to use anyOf in place of oneOf for the `interactionType` property.  Since it could be a string enum or a string, validation could not differentiate.
* Fix `fdc3.timeRange` context example to use correctly formatted dateTime.

### Related Issues
Resolves #1592
Resolves #1598 
Resolves #1599

## Contributor License Agreement

<!--- All contributions to FDC3 must be made under an active contributor license agreement and the [Community Specification License](https://github.com/finos/FDC3/blob/main/LICENSE.md). This will be checked by the EasyCLA tool (https://easycla.lfx.linuxfoundation.org/) that runs automatically on every PR. If you've not contributed to FDC3 before, look for a comment on your PR shortly after it is raised and follow the instructions to establish a CLA or have it acknowledged by the EasyCLA tool. -->

- [X] I acknowledge that a contributor license agreement is required and that I have one in place or will seek to put one in place ASAP.

## Review Checklist

<!--- Checklist to be completed by reviewers, and pre-checked by the authors of a PR -->

- [X] **Issue**: If a change was made to the FDC3 Standard, was an issue linked above?
- [X] **CHANGELOG**: Is a *CHANGELOG.md* entry included?
- [ ] **API changes**: Does this PR include changes to any of the FDC3 APIs (`DesktopAgent`, `Channel`, `PrivateChannel`, `Listener`, `Bridging`)?
  - [ ] **Docs & Sources**: If yes, were both documentation (/docs) and sources updated?<br/>
        *JSDoc comments on interfaces and types should be matched to the main documentation in /docs*
  - [ ] **Conformance tests**: If yes, are conformance test definitions (/toolbox/fdc3-conformance) still correct and complete?<br/>
        *Conformance test definitions should cover all **required** aspects of an FDC3 Desktop Agent implementation, which are usually marked with a **MUST** keyword, and  optional features (**SHOULD** or **MAY**) where the format of those features is defined*
  - [X] **Schemas**: If yes, were changes applied to the Bridging and FDC3 for Web protocol schemas?<br/>
        *The Web Connection protocol and Desktop Agent Communication Protocol schemas must be able to support all necessary aspects of the Desktop Agent API, while Bridging must support those aspects necessary for Desktop Agents to communicate with each other*
    - [X] If yes, was code generation (`npm run build`) run and the results checked in?<br/>
        *Generated code will be found at `/src/api/BrowserTypes.ts` and/or `/src/bridging/BridgingTypes.ts`*
- [x] **Context types**: Were new Context type schemas created or modified in this PR?
  - [ ] Were the [field type conventions](https://fdc3.finos.org/docs/context/spec#field-type-conventions) adhered to?
  - [ ] Was the `BaseContext` schema applied via `allOf` (as it is in existing types)?
  - [ ] Was a `title` and `description` provided for all properties defined in the schema?
  - [x] Was at least one example provided?
  - [x] Was code generation (`npm run build`) run and the results checked in?<br/>
        *Generated code will be found at `/src/context/ContextTypes.ts`*
- [ ] **Intents**: Were new Intents created in this PR?
  - [ ] Were the [intent name prefixes](https://fdc3.finos.org/docs/intents/spec#intent-name-prefixes) and other [naming conventions & characteristics](https://fdc3.finos.org/docs/intents/spec#naming-conventions) adhered to?
  - [ ] Was the new intent added to the list in the [Intents Overview](https://fdc3.finos.org/docs/intents/spec#standard-intents)?


---
THIS SOFTWARE IS CONTRIBUTED SUBJECT TO THE TERMS OF THE FINOS CORPORATE CONTRIBUTOR LICENSE AGREEMENT.
 
THIS SOFTWARE IS LICENSED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE AND ANY WARRANTY OF NON-INFRINGEMENT, ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. THIS SOFTWARE MAY BE REDISTRIBUTED TO OTHERS ONLY BY EFFECTIVELY USING THIS OR ANOTHER EQUIVALENT DISCLAIMER IN ADDITION TO ANY OTHER REQUIRED LICENSE TERMS.

kriswest reviewed (approved):
LGTM thanks @bingenito 

@finos/fdc3-editors could we get a review from one of you?

paulgoldsmith commented on CHANGELOG.md:
```diff
@@ -14,6 +14,13 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
 ### Fixed
 
+## [npm v2.2.1] - 2025-05-16
```
```suggestion
## [npm v2.2.1] - 2025-05-23
```

kriswest commented:
Consent was requested from SWG on 22nd May 2025 and was given

bingenito commented:
@kriswest and @paulgoldsmith I've changed this quite a bit since last review.  I've added the tests and after creating the tests identified the same issue in fdc3.Interaction as well as an invalid datetime in the timeRange example.  The example change and the fdc3.Interaction change did result in a doc change so both next (trunk) and 2.2 were updated.  I've also reverted the package version change and moved the changelog back into unreleased as tying it down was making it hard to schedule and review the PR.

bingenito commented:
If you look at the status check (which is failing because of fork pr permissions) you can see the output of the new tests.  It will also fail if any context is added that doesn't include an example to validate.

```
> @finos/fdc3-context@2.2.0 lint
> eslint generated/ --fix && npx prettier generated/ --write

generated/context/ContextTypes.ts 411ms
PASS test/validate-schema-examples.test.ts
  FDC3 Context Schema Example Validation
    ✓ action.schema.json example #1 should be valid (2 ms)
    ✓ action.schema.json example #2 should be valid (1 ms)
    ✓ chart.schema.json example #1 should be valid (1 ms)
    ✓ chatInitSettings.schema.json example #1 should be valid (4 ms)
    ✓ chatMessage.schema.json example #1 should be valid (2 ms)
    ✓ chatRoom.schema.json example #1 should be valid
    ✓ chatSearchCriteria.schema.json example #1 should be valid
    ✓ contact.schema.json example #1 should be valid
    ✓ contactList.schema.json example #1 should be valid
    ✓ country.schema.json example #1 should be valid (1 ms)
    ✓ currency.schema.json example #1 should be valid (1 ms)
    ✓ email.schema.json example #1 should be valid
    ✓ fileAttachment.schema.json example #1 should be valid
    ✓ instrument.schema.json example #1 should be valid
    ✓ instrumentList.schema.json example #1 should be valid
    ✓ interaction.schema.json example #1 should be valid
    ✓ message.schema.json example #1 should be valid
    ✓ nothing.schema.json example #1 should be valid
    ✓ order.schema.json example #1 should be valid
    ✓ order.schema.json example #2 should be valid
    ✓ orderList.schema.json example #1 should be valid
    ✓ organization.schema.json example #1 should be valid
    ✓ portfolio.schema.json example #1 should be valid (1 ms)
    ✓ position.schema.json example #1 should be valid
    ✓ product.schema.json example #1 should be valid (1 ms)
    ✓ timeRange.schema.json example #1 should be valid
    ✓ timeRange.schema.json example #2 should be valid
    ✓ timeRange.schema.json example #3 should be valid
    ✓ trade.schema.json example #1 should be valid (1 ms)
    ✓ tradeList.schema.json example #1 should be valid
    ✓ transactionresult.schema.json example #1 should be valid (1 ms)
    ✓ valuation.schema.json example #1 should be valid (1 ms)

Test Suites: 1 passed, 1 total
Tests:       32 passed, 32 total
Snapshots:   0 total
Time:        0.6[87](https://github.com/finos/FDC3/actions/runs/15322023353/job/43107714097?pr=1593#step:7:88) s
Ran all test suites.
```

kriswest commented:
Looks like a great addition @bingenito, thank you! I'm glad you are failing out when no example is present.

I'd try and apply it to the FDC3 for Web schemas, but that would mean writing *a lot* of examples, which will be less use there anyway.

Looks like you're failing the SAST scan [here](https://github.com/bingenito/FDC3/blob/936cbe279f918120b42f0d4b66c041d4fabad7d7/packages/fdc3-context/test/validate-schema-examples.test.ts#L55) and [here](https://github.com/bingenito/FDC3/blob/936cbe279f918120b42f0d4b66c041d4fabad7d7/packages/fdc3-context/test/validate-schema-examples.test.ts#L66) and it think syou've possibly got unsanitized user input going into `path.join` (which you don't - the rule is just not smart enough to work out that its a fixed list of files at a location in the project. 

To resolve you can either just add an exception above each line, e.g:

```
// nosemgrep: javascript.lang.security.audit.path-traversal.path-join-resolve-traversal.path-join-resolve-traversal
const schemaPath = path.join(apiSchemasDir, schemaFile);
```

or wrap the filename in a sanitizer function (so it doesn't think the input is being passed directly to `path.join`). Here is one from https://semgrep.dev/r?q=javascript.lang.security.audit.path-traversal.path-join-resolve-traversal.path-join-resolve-traversal

```js
function okTest3(req,res) {
    let somePath = req.body.path;
    somePath = somePath.replace(/^(\.\.(\/|\\|$))+/, '');
    // ok:path-join-resolve-traversal
    return path.join(opts.path, somePath);
}
```

having written that out, I might as well add the suggestions...

kriswest commented on packages/fdc3-context/test/validate-schema-examples.test.ts:
```diff
@@ -0,0 +1,112 @@
+import path from 'path';
+import Ajv from 'ajv';
+import addFormats from 'ajv-formats';
+
```
Filename sanitizing fn needed below to past SAST scans:

```suggestion
const sanitizeFilename = (filename: string) => filename.replace(/^(\.\.(\/|\\|$))+/, '');
```

kriswest commented on packages/fdc3-context/test/validate-schema-examples.test.ts:
```diff
@@ -0,0 +1,112 @@
+  if (fs.existsSync(apiSchemasDir)) {
+    const apiSchemaFiles = fs.readdirSync(apiSchemasDir).filter(f => f.endsWith('.schema.json'));
+    apiSchemas = apiSchemaFiles.map(schemaFile => {
+      const schemaPath = path.join(apiSchemasDir, schemaFile);
```
```suggestion
      const schemaPath = path.join(apiSchemasDir, sanitizeFilename (schemaFile));
```

kriswest commented on packages/fdc3-context/test/validate-schema-examples.test.ts:
```diff
@@ -0,0 +1,112 @@
+
+  // Load and map all context schemas
+  const schemas = schemaFiles.map(schemaFile => {
+    const schemaPath = path.join(schemasDir, schemaFile);
```
```suggestion
    const schemaPath = path.join(schemasDir, sanitizeFilename (schemaFile));
```

kriswest reviewed (changes requested):
Needs to pass SAST scans - see comments inline.

GREAT addition @bingenito!

kriswest commented:
@hughtroeger and @novavi could you take a look at this PR please

kriswest commented:
Rerunning the build job due to a weird failure: https://github.com/finos/FDC3/actions/runs/15335055464/job/43150484515?pr=1593
